crates/wasmtime/src/runtime/component/func/host.rs

```diff
@@ -225,20 +294,42 @@ where
             Storage::Indirect(slice_to_storage_mut(storage).assume_init_ref())
         }
     };
-    let mut lift = LiftContext::new(cx.0, &options, types, instance);
-    lift.enter_call();
-    let params = storage.lift_params(&mut lift, param_tys)?;
 
-    let ret = closure(cx.as_context_mut(), params)?;
+    let replay_enabled = cx.0.replay_enabled();
+    let ret = if !replay_enabled {
+        ReturnMode::Standard({
+            let mut lift = LiftContext::new(cx.0, &options, types, instance);
+            lift.enter_call();
+
+            let params = storage.lift_params(&mut lift, param_tys)?;
+            closure(cx.as_context_mut(), params)?
+        })
+    } else {
+        ReturnMode::Replay
+    };
+
     flags.set_may_leave(false);
     let mut lower = LowerContext::new(cx, &options, types, instance);
-    storage.lower_results(&mut lower, result_tys, ret)?;
+    storage.lower_results(
+        &mut lower,
+        result_tys,
+        ret,
+        #[cfg(feature = "rr-type-validation")]
+        &types[ty.results],
+    )?;
     flags.set_may_leave(true);
 
-    lower.exit_call()?;
+    if !replay_enabled {
+        lower.exit_call()?;
+    }
 
     return Ok(());
 
+    enum ReturnMode<Return: Lower> {
```

Comment:
Can we add a doc-comment here describing why we need a separate `Replay` return-mode? (Basically, the effects of the return are handled by replaying the trace instead, right?)